packages/react/src/combobox/item/ComboboxItem.tsx

```diff
@@ -0,0 +1,271 @@
+  React.forwardRef(function ComboboxItem(
+    componentProps: ComboboxItem.Props,
+    forwardedRef: React.ForwardedRef<HTMLDivElement>,
+  ) {
+    const {
+      render,
+      className,
+      value,
+      disabled = false,
+      nativeButton = false,
+      ...elementProps
+    } = componentProps;
+
+    const id = useBaseUiId();
+
+    const textRef = React.useRef<HTMLElement | null>(null);
+    const listItem = useCompositeListItem({
+      textRef,
+      indexGuessBehavior: IndexGuessBehavior.GuessFromOrder,
+    });
+
+    const {
+      store,
+      selectionMode,
+      getItemProps,
+      setOpen,
+      setSelectedValue,
+      valuesRef,
+      inputRef,
+      registerItemIndex,
+      allowActiveIndexSyncRef,
+      onItemHighlighted,
+      readOnly,
+      virtualized,
+      listRef,
+    } = useComboboxRootContext();
+
+    const { flatFilteredItems } = useComboboxDerivedItemsContext();
+
+    const selectable = selectionMode !== 'none';
+    const multiple = selectionMode === 'multiple';
+    const index = virtualized ? flatFilteredItems.indexOf(value) : listItem.index;
+
+    const isRow = useComboboxRowContext();
+    const active = useStore(store, selectors.isActive, index);
+    const matchesSelectedValue = useStore(store, selectors.isSelected, value);
+    const rootSelectedValue = useStore(store, selectors.selectedValue);
+    const initialList = useStore(store, selectors.initialList);
+
+    const selected = matchesSelectedValue && selectable;
+
+    const itemRef = React.useRef<HTMLDivElement | null>(null);
+    const indexRef = useLatestRef(index);
+
+    const frame = useAnimationFrame();
+
+    const hasRegistered = listItem.index !== -1;
+
+    useIsoLayoutEffect(() => {
+      if (!virtualized) {
+        return undefined;
+      }
+
+      const list = listRef.current;
+      const values = valuesRef.current;
+
+      if (index !== -1) {
+        list[index] = itemRef.current;
+        // Keep valuesRef aligned with the filtered list indices in virtualized mode
+        values[index] = value;
+        return () => {
+          delete list[index];
+          delete values[index];
+        };
+      }
+
+      return undefined;
+    }, [index, virtualized, listRef, valuesRef, value, initialList]);
+
+    useIsoLayoutEffect(() => {
+      if (!hasRegistered) {
+        return undefined;
+      }
+
+      const values = valuesRef.current;
+      values[index] = value;
+
+      return () => {
+        delete values[index];
+      };
+    }, [hasRegistered, index, value, valuesRef]);
+
+    useIsoLayoutEffect(() => {
+      if (hasRegistered && value === rootSelectedValue) {
+        registerItemIndex(index);
+
+        if (selectable && allowActiveIndexSyncRef.current) {
+          frame.request(() => {
+            itemRef.current?.scrollIntoView?.({ inline: 'nearest', block: 'nearest' });
+            store.set('activeIndex', index);
+            store.set('selectedIndex', index);
```

Comment:
The store is currently synchronous by design, so state updates are not batched. It's more efficient to update the state in one call with `store.apply()` (that I'll refactor to `store.update()` in the near future). Eventually we might have opt-in explicit batching (e.g. `store.batch(() => { ... })`), but for now we don't because it makes state updates predictable (as soon as you call it, you know all the listeners have been called).